sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogV2Util.scala

```diff
@@ -598,10 +598,28 @@ private[sql] object CatalogV2Util {
       //       data unchanged and let the data reader to return "exist default" for missing
       //       columns.
       val existingDefault = Literal(default.getValue.value(), default.getValue.dataType()).sql
-      f.withExistenceDefaultValue(existingDefault).withCurrentDefaultValue(default.getSql)
+      f.withExistenceDefaultValue(existingDefault).withCurrentDefaultValue(toSql(defaultValue))
     }.getOrElse(f)
   }
 
+  private def toSql(defaultValue: DefaultValue): String = {
+    if (defaultValue.getExpression != null) {
```

Comment:
iiuc, this is prioritizing expression over sql , if we get both.  Is that desired?  We could put a comment somewhere if so

Reply:
Yea I have the same question. It seems more reasonable to use the original SQL string first.

Reply:
Based on `DefaultValue` document, it uses expression first and fallbacks to SQL string if the expression cannot be converted.

Reply:
When we need to evaluate, it makes sense to use expression first and fallback to SQL string if the expression cannot be converted. But here we just need to get the SQL string.

Reply:
The problem is that whatever we encode here will be picked up and used on write. Spark doesn't currently use expressions to represent default values. We persist the expression SQL in struct field metadata. What this logic does is saying if I a connector provides an expression, I am going to convert it to Catalyst and trust Spark to generate an accurate SQL representation. If we flip this, it means we prioritize the connector-provided SQL string over our own SQL generation for Catalyst expressions.

Reply:
Does this make sense, @szehon-ho @viirya @cloud-fan?

Reply:
But Spark is not capable of generating an executable SQL string from a query plan/expression. The `.sql` only generates a pretty string for display. We don't even have a test to parse back the `.sql` result.

On the other hand, the `DefaultValue.sql` is usually the original user-specified SQL, or the connector is responsible to give an executable SQL if it generates it by itself.

Reply:
Ah, I misinterpreted the promise behind `.sql`. It was my understanding it is actually reliable. Let me think more about this tomorrow. If `.sql` is not reliable, then it does change things a bit, I guess. The problem that if we prefer the string over the expression it kind of breaks the promise behind `DefaultValue`.

Reply:
There is a way to produce reliable SQL strings from v2 expressions: `V2ExpressionSQLBuilder`

Reply:
@cloud-fan, that could be an option. However, it doesn't produce Spark SQL dialect... for things like DAY_OF_WEEK and similar expressions.

The fundamental problem is that we have a Catalyst expression and we need to persist it in the struct field metadata. I wonder if we can avoid the need to put the expression in the metadata. It is going to be challenging for sure, as it requires changing `StructField`, which is a very stable API.

Any thoughts?

Reply:
If it's just for the legacy v1 code path (e.g. built-in file sources), I'd say let's keep the old behavior and save the original SQL string only. We can refactor them to use `ColumnDefinition` instead of `StructType`, but that is non trival and we can leave it for the future.